Xtensa asm2 framework

andyross opened the pull request:
Second cut.  This one is ready for merge if anyone feels it's ready.   We can hold off though too: the integration bits are another day or two out, so this is just the primitives and unit tests

Commit:
xtensa: Move register window exception handlers into a separate file

No behavior changes, just code motion.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer/xtensa_sys_timer: Correctly declare ISR

The existing hand-written interrupt code is manually calling the timer
ISR, which is sort of silly and about to be replaced.  Correctly
declare the ISR with IRQ_CONNECT() so that a conventional interrupt
handling implementation can find it.  With current code this is a
noop.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Make high priority interrupts optional

Xtensa has a "high priority" class of interrupt levels which ignore
the EXCM bit and can thus interrupt running exception handlers.  These
can't be used for C handlers in the general case[1] because C code
needs to be able to throw window over/underflow exceptions, which are
not reentrant.

But the high priority interrupts might be useful to a carefully
designed application, or to unit tests of low level architecture code.
So make their generation optional with this kconfig option.

[1] ESP-32 has a high priority interrupt for its watchdog, apparently.
    Which is sort of OK given that it never needs to return to the
    interrupted code.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: New asm layer to support SMP

SMP needs a new context switch primitive (to disentangle _swap() from
the scheduler) and new interrupt entry behavior (to be able to take a
global spinlock on behalf of legacy drivers).  The existing code is
very obtuse, and working with it led me down a long path of "this
would be so much better if..."  So this is a new context and entry
framework, intended to replace the code that exists now, at least on
SMP platforms.

New features:

* The new context switch primitive is xtensa_switch(), which takes a
  "new" context handle as an argument instead of getting it from the
  scheduler, returns an "old" context handle through a pointer
  (e.g. to save it to the old thread context), and restores the lock
  state(PS register) exactly as it is at entry instead of taking it as
  an argument.

* The register spill code understands wrap-around register windows and
  can avoid spilling A4-A15 registers when they are unused by the
  interrupted function, saving as much as 48 bytes of stack space on
  the interrupted stacks.

* The "spill register windows" routine is entirely different, using a
  different mechanism, and is MUCH FASTER (to the tune of almost 200
  cycles).  See notes in comments.

* Even better, interrupt entry can be done via a clever "cross stack
  call" I worked up, meaning that the interrupted thread's registers
  do not need to be spilled at all until they are naturally pushed out
  by the interrupt handler or until we return from the interrupt into
  a different thread.  This is a big efficiency win for tiny
  interrupts (e.g. timers), and a big latency win for all interrupts.

* Interrupt entry is 100% symmetric with respect to medium/high
  interrupts, avoiding the problems seen with hooking high priority
  interrupts with the current code (e.g. ESP-32's watchdog driver).

* Much smaller code size.  No cut and paste assembly.  No use of HAL
  calls.

* Assumes "XEA2" interrupt architecture, the register window extension
  (i.e. no CALL0 ABI), and the "high priority interrupts" extension.
  Does not support the legacy processor variants for which we have no
  targets.  The old code has some stuff in there to support this, but
  it seems bitrotten, untestable, and I'm all but certain it doesn't
  work.

Note that this simply adds the primitives to the existing tree in a
form where they can be unit tested.  It does not replace the existing
interrupt/exception handling or _Swap() implementation.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
samples/xtensa-asm2: Unit test for new Xtensa assembly primitives

This sample (which should eventually become a proper test) suite
builds from simple applications of the new primitives to a full
context switch test and interrupt handling suite (based on the
CPU-internal CCOMPARE2 timer).

It's been extraordinarily useful finding regressing as the asm2 code
gets modified and should probably stick around as long as possible.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Forgot to note: I cleaned up the true positive checkpatch issues.  The remaining warnings are all due to assembly syntax.

Commit:
kernel: Fix nano_internal.h inclusion

_Swap() is defined in nano_internal.h.  Everything calls _Swap().
Pretty much nothing that called _Swap() included nano_internal.h,
expecting it to be picked up automatically through other headers (as
it happened, from the kernel arch-specific include file).  A new
_Swap() is going to need some other symbols in the inline definition,
so I needed to break that cycle.  Now nothing sees _Swap() defined
anymore.  Put nano_internal.h everywhere it's needed.

Our kernel includes remain a big awful yucky mess.  This makes things
more correct but no less ugly.  Needs cleanup.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Move some macros

K_NUM_PRIORITIES and K_NUM_PRIO_BITMAPS were defined in
nano_internal.h, but used in only a handful of places.  Move to
kernel_structs.h (somewhat higher up in the hierarchy) to help with
include file cycle-breaking.  Arguably they are a better fit there
anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Remove _get_next_ready_thread() from sched API

This function was never used outside of sched.c, so move it there.
With SMP coming, it's a racy API (the other CPU can schedule that
"next" thread at any time!) that we don't want exposed.

(And no, I don't know what this was doing as an unused declaration in
a RISC-V assembly file either)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Add alternative _arch_switch context switch primitive

The existing __swap() mechanism is too high level for some
applications because of its scheduler-awareness.  This introduces a
new _arch_switch() mechanism, which is a simpler primitive that looks
like:

    void _arch_switch(void *handle, void **old_handle_out);

The new thread handle (typically just a stack pointer) is specified
explicitly instead of being picked up from the scheduler by
per-architecture code, and on return the "old" thread handle that got
switched out is returned through the pointer.

The new primitive (currently available only on xtensa) is selected
when CONFIG_USE_SWITCH is "y".  A new C _Swap() implementation based
on this primitive is then added which operates compatibly.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa/asm2: Add a _new_thread implementation for asm2/switch

Implement _new_thread in terms of the asm2 switch mechanism.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Add Kconfig for asm2 layer

The asm2 layer will build alongside the traditional assembly, but the
reverse is not true.  Add a CONFIG_XTENSA_ASM2 to force its use at
runtime and disable the older code.

Note that the older assembly had an initialization function that is
properly part of the timer driver.  Move a C equivalent into the timer
driver itself for now to prevent a build breakage.  Long term we need
to clean that driver up in a bunch of other ways.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Add first phase of integration plumbing.  hello_world now runs correctly.  Interrupts are next.

Commit:
xtensa: Remove arch-specific (and empty) offsets.h

The xtensa arch code had this empty offsets.h header sitting around.
Its name collides with the autogenerated offsets.h, making it
dangerously dependent on include file path order.  Seems to be benign,
but it's freaking me out.  Remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented on drivers/timer/xtensa_sys_timer.c:
```diff
@@ -490,6 +504,8 @@ void _timer_int_handler(void *params)
  */
 int _sys_clock_driver_init(struct device *device)
 {
+	IRQ_CONNECT(TIMER_IRQ, 0, _timer_int_handler, 0, 0);
```
huzzah!

andrewboie commented on samples/xtensa_asm2/src/main.c:
```diff
@@ -0,0 +1,539 @@
+				    (void *)test_switch_bounce,
+				    0, 0, 0);
+
+#if 0
```
consider getting rid of this unused code block, or conditionally compiling with some debug Kconfig instead of #if 0.
If you really want it set up this way it's fine w/me, more of a suggestion.

nashif replied:
this code block should be removed.

andrewboie commented on samples/xtensa_asm2/src/main.c:
```diff
@@ -0,0 +1,539 @@
+
+		enable_timer();
+
+#if 0
```
same as above

andrewboie commented on samples/xtensa_asm2/src/main.c:
```diff
@@ -0,0 +1,539 @@
+
+	irq_unlock(key);
+
+	printk("%s\n", ok ? "OK" : "Failed");
```
please convert this to ztest so that success/failure is reported in the expected way and move this under tests/.
We want sanitycheck/test case framework to exercise this in qemu-xtensa or real hardware, since we appear to be able to determine success/failure at runtime.

nashif replied:
agree

tyrantlizard77 reviewed (commented):
A lot of the ASM stuff is very obscure to me. I would be more comfortable about its correctness if we were exercising it in CI, which we currently aren't.

I am not super into the idea of having two parallel implementations of the asm domain for Xtensa. I think that since we are early in the 1.11 cycle we should just commit to the new one, so that it gets exercised in test case framework and sanitycheck.

I also must insist on the provided sample to be converted to a ztest case, or at least converted to an old-style test, such that it gets run by sanitycheck and reports its success/failure in a way that our test case infra understands.

andyross commented:
Agreed that the test integration is needed.  Though this is going to need a stabilization phase as always.  So far it's been hugely useful to be able to run the primitives underneath the existing framework as unit tests, and that will probably continue to be true for a bit.  I'd like to keep it around at least through one release.

But sure: at the end of the process, and if we're OK with dropping the "dead" platforms in there (i.e. everything but qemu and ESP-32, our only supported board targets), then sure: I'd be happen to push a patch removing it all.

Commit:
xtensa: Add exception/interrupt vectors in asm2 mode

This adds vectors for all interrupt levels defined by core-isa.h.

Modify the entry code a little bit to select correct linker sections
(levels 1, 6 and 7 get special names for... no particularly good
reason) and to constructed the interrupted PS value correctly (no EPS1
register for exceptions since they had to have interrupted level 0
code and thus differ only in the EXCM bit).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Interrupt generator script and output for qemu & esp32

This python script reads the core-isa.h interrupt defintions (via
running a template file through the toolchain preprocessor to generate
an input file) and emits a fully populated, optimized C handling code
that binary searches only the declared interrupts at a given level and
correctly detects spurious interrupts (and/or incorrect core-isa.h
definitions).

The generated code, alas, turns out not to be any faster than simply
searching the interrupt mask with CLZ (er, NSAU in xtensese), though
it could be faster in theory if the compiler made different choices,
see comments.  But I like this for the robustness of the fully
populated search trees and the checking of level vs. mask.

This simply commits the script output into the source tree, including
some checking code to force a build error if the toolchain changes the
headers incompatibly.  It would be better long term to have these
headers be generated at build time, but that requires more cmake fu
than I have.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Remove _xt_set_exception_handler()

This was a dead API.  Nothing ever used it, it wasn't exposed in any
API headers.  It never appeared in documentation.  It's not
particularly clear why a Zephy app would want to hook
architecture-specific exceptions instead of simply using the portable
error framework anyway. And it's not supported by asm2.  Delete.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Implement _xt_ints_on/off for asm2

Legacy xtensa had a rather complicated implementation of en/disabling
interrupts, owing to the "software priority" feature (which plays
games with INTENABLE and INTLEVEL to allow for interrupts to interrupt
each other outside their normal priorities).  But that's not a Zephyr
feature, it's enabled by a XT_USE_SWPRI value that comes from platform
headers and isn't enabled on any of our boards.  Dead code, basically.

Replace with the obvious implementation when asm2 is in use.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer/xtensa_sys_timer: Add init/update hooks for asm2

The earlier xtensa layer put the timer initialization and update
directly into the interrupt handler, which is... weird.  Under asm2,
it's just a regular ISR and needs to do the work in the driver.

Really, this driver needs a bunch of cleanup.  The xtensa CPU timer is
two registers and one ISR: a global cycle count register, and a
compare register that will fire the IRQ when they match.  There is
*way* too much code here.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Move _get_next_ready_thread() to the scheduler API

The old __swap implementations would get at this via a struct offset
in assembly code, but _arch_switch code needs it in C.  Just move it
to the header.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Next cut.  This is 100% complete now and passes all my own unit tests.  Runs a bunch of the broader tests (when setting the CONFIG_XTENSA_ASM2=y as a default), still fails on a few.

lpereira commented on arch/xtensa/soc/esp32/include/_soc_inthandlers.h:
```diff
@@ -0,0 +1,352 @@
+
+static inline int _xtensa_handle_one_int1(unsigned int mask)
+{
+	if (mask & 0x7f) {
```
I wonder why this is needed; can't something that uses `__builtin_ffs()` be used instead? This code must be in the ESP32 IRAM and compete with other things that must be there (compared with things that can live in flash and be used via the flash cache).

```C
int irq = _builtin_ffs(mask);
if (irq) {
        _sw_isr_table[irq].isr(_sw_isr_table[irq].arg);
       return 1<<irq;
}
return 0;
```

andyross commented:
Regarding __builtin_ffs: the builtin is actually slower.  GCC emits code to call a function and roll the register windows instead of putting it inline.  There is a (technically optional, probably why GCC doesn't inline it) NSAU instruction (not quite exactly CLZ/FFS, but same general idea) that could be used.  But... if GCC would generate optimal code, this trick (binary searching the masks) would actually beat that too.  But it doesn't, because it won't emit a BBS instruction (test a static bit and branch in a single instruction) and instead wants to test a mask for the single bits, which defeats the gain.

So... yeah, some rework here for cycle optimization couldn't hurt.  Generating assembly for the selection would be an option.  Another trick would be to integrate it with the ISR table generation so you can have the static function pointer and argument inline in the code instead of loaded from memory.  That's another two cycles.

andyross commented:
Oh, and I realized I didn't answer your question: that code is the basic idea, but you also have the problem that the INTERRUPT register will flag interrupts from all levels, not just the one your are servicing.  So you need to make sure that a level N interrupt checks and calls only functions from level N, or else you might accidentally call a low-priority ISR out of a high priority handler simply because they fired at the same time.  (Which is why I wanted to generate this code from core-isa.h in the first place).

lpereira commented:
My main issue isn't much with interrupt latency -- but with code size.  At least on the ESP32, IRAM is quite limited (32KiB IIRC) and the whole kernel, drivers, and ISRs must be placed there. The reason being that that port currently runs entirely off IRAM (it's placed there by the 1st stage bootloader); and, when the flash cache is enabled, code that handles IRQs must be available. If some code must be paged in to the flash cache when an interrupt is being handled, a fault will occur.

This isn't bad (Zephyr is small enough that it fits comfortably in IRAM), but anything we can do to reduce the code size for something that's not optional is welcome.

andyross commented:
Gotcha.  Yeah, a tighter loop wouldn't hurt.  To be fair: the last I checked this series as a whole knocked like 1100 bytes off the .text size, so there's a little room for waste.

lpereira commented:
Yeah, 1KB of code is indeed a lot.  If anything, I'm fine with the way things are, given all the improvements made in this series.  It's just something that we can revisit in the future if space is at a premium.

andrewboie commented on arch/xtensa/Kconfig:
```diff
@@ -76,6 +76,35 @@ config IRQ_OFFLOAD_INTNUM
+config XTENSA_ASM2
+	bool
+	prompt "New-sytle Xtensa context switch & interrupt layer"
+	default n
```
I would like to see this default Y unless it breaks something

andrewboie commented:
> dropping the "dead" platforms

Not sure I understand this.
Does the new code not work with the various xt-sim targets?

andyross commented:
It only probably works with the xcc targets (at least assuming they are all windowed variants -- without register windows this definitely doesn't work).  Frankly I don't even have a way to enumerate the core-isa.h files for those to find out; they're s all buried behind the proprietary SDK.  As it stands it definitely supports only qemu_xtensa and ESP-32.

Again, the goal here was to have a reworked interrupt and swap framework for SMP, and it grew from there.  I'd love for this to eventually eat the world, but IMHO it's fine to have a "legacy" layer for the old/proprietary targets with vendor support and a more modern one for current work.

andrewboie commented:
> I'd love for this to eventually eat the world, but IMHO it's fine to have a "legacy" layer for the old/proprietary targets with vendor support and a more modern one for current work.

Works for me

andyross commented on arch/xtensa/include/xtensa-asm2-s.h:
```diff
@@ -297,26 +319,43 @@ _restore_\@:
+	/* Level "1" is the exception handler, which uses a different
+         * calling convention.  No special register holds the
+         * interrupted PS, instead we just assume that the CPU has
+         * turned on the EXCM bit and set INTLEVEL.
```
This comment is wrong, "and set INTLEVEL" is actually not what it does (though it's what it *should* do, sigh).  Thus the "gotcha" comment and code above.  Will fix.

andyross reviewed (commented):
Hm... reading through there's some stuff I have to fix before this merges.

And I'm looking at what is for sure a real bug with switching a brand new stack early (or something) that the Dining Philosophers test is triggering, and that needs to be fixed.

Hold off merging for now.  Will try to fix from New Zealand.

andyross commented on arch/xtensa/core/xtensa_intgen.py:
```diff
@@ -0,0 +1,130 @@
+import re
+import fileinput
+
+# Pass an Xtensa core-isa.h file on stdin or the command line, emits a
```
Another lying comment.  It takes the output of preprocessing the template file, not core-isa.h itself.

lpereira commented on arch/xtensa/core/xtensa_intgen.py:
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
```
Just a nit, but this script needs a copyright+license header.

lpereira commented on arch/xtensa/core/xtensa_intgen.py:
```diff
@@ -0,0 +1,130 @@
+# Populate empty levels just for sanity.  The second-to-last interrupt
+# level (usually "debug") typically doesn't have any associated
+# vectors, but we don't have any way to know that a-prioi.
+max = 0
```
Python has a max() builtin that works pretty much with everything: you can pass various arguments or any iterable object. So this loop can be written as `max_level = max(ints_by_lvl)`.

lpereira commented on arch/xtensa/core/xtensa_intgen.py:
```diff
@@ -0,0 +1,130 @@
+    cprint("{")
+
+    ints_by_lvl[lvl].sort()
+    emit_int_handler(ints_by_lvl[lvl])
```
Not really necessary, but you can simplify this slightly by removing the previous `for lvl...` loop, and writing it like this: `emit_int_handler(sorted(ints_by_lvl.get(lvl, [])))`.  Maybe even remove the `max_level = max(...)` line as well.

andyross replied:
Yeah.  I write toddler python.  Guilty as charged.

Commit:
kernel, xtensa: Switch-specific thread return value

When using _arch_switch() context switching, the thread return value
is a generic hook and not provided by the architecture.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Fix thread entry point

The stack initilaization was calling the user-provided entry function
directly, which works fine until that function returns, at which point
it will try to unspill A0-A3 from the 16 bytes above the allocated
stack and then "return" to a NULL pointer.

The kernel provides a _thread_entry() function that does cleanup
properly, so use that.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: asm2 needs to honor thread preemption

Forgot to check the thread preemption status when fetching the stack
to restore.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: THREAD_MONITOR hooks for asm2

You'd this feature would be portable, but it's arch-specific.
Initialize the CONFIG_THREAD_MONITOR stuff, placing the __thread_entry
struct (which AFAICT is dead: nothing in the tree actually reads it)
at the top of the stack.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa asm2: Fixup stack alignment at runtime

The API allows any byte count for stack size, and tests in fact check
that a stack with a 499 byte stack works correctly.  No choice, have
to do this at runtime.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

nashif commented on samples/xtensa_asm2/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+set(IS_TEST 1)
```
move to samples/boards

nashif replied:
and this is more of a test than a sample, can we move this to tests/ please?

Commit:
xtensa: Add hook to do register window spills

This macro was already available add an external symbol so C code can
access it (via CALL0 -- it's not and can't be an actual function).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa-asm2: Exception/interrupt handler should check stack sentinel

This got forgotten.  Note that this function is empty if
CONFIG_STACK_SENTINEL is unconfigured.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa-asm2: Handle alloca/movsp exceptions

Xtensa register windows have a special exception that happens when the
stack pointer needs to be moved, but the caller function has already
spilled its registers below it.

I thought these were unexercised in Zephyr code, but they turn out to
be thrown by the existing mem_pool tests when run in the 32-register
qemu environment (but not on 64-register hardwre).  Because the effect
of the exception is to unspill the caller, there is no good way to
handle this in a traditional handler.  Instead put a 5-instruction
stub in front of the user exception handler (i.e. incurring that cost
on every trap and every L1 interrupt) to test before doing the normal
entry.

Works, but would be nicer to optimize this in the future so that only
true alloca exceptions take that cost.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Dust this off and push all the final fixes.  Finally got all of sanitycheck running on qemu_xtensa and a decent amount of breadth on manual esp32 testing.  Switching over to asm2 exclusively on those platforms, so the CI test will show a proper result using the new code.

I do see the comments on the sample/test code which is still the original.  Will address, though maybe it would be easier to remove that from the series if someone wants to merge immediately.

nashif commented:
@rgundi can you please test this patch set with Intel S1000 platform?

andyross commented:
Note that Xtensa SoCs other than ESP-32 and qemu (which is called "sample_controller" in our tree for reasons I don't understand) will still build with the legacy code.  That absolutely should be tested and shouldn't break.  But we'll want to migrate other supported platforms ASAP.   Requires running the one python script to generate the interrupt stubs against the per-SoC toolchain.  Bug me offline.

andrewboie commented:
> which is called "sample_controller" in our tree for reasons I don't understand

The SOC is called this because the specific CPU build in the Cadence SDK it is targeting is also named "sample_controller".

Commit:
esp-32, qemu_xtensa: Use asm2 by default

Set these SoCs to use asm2 by default

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
And of course I check this morning and discover I pushed an unrebased tree.  Now it should work.

nashif commented:
can we close this?
